runtime/ast/unmarshal.go

```diff
@@ -0,0 +1,901 @@
+type TypeUnmarshaler struct{ Type *Type }
+
+func (a *BigIntUnmarshaler) UnmarshalJSON(b []byte) (err error) {
+	s := string(b)
+	if s, err = strconv.Unquote(s); err != nil {
+		return err
+	} else if s == "null" {
+		*a.Int = nil
+		return nil
+	}
+	if z, ok := (&big.Int{}).SetString(s, 10); !ok {
+		return fmt.Errorf("invalid BigInt %s", s)
+	} else {
+		*a.Int = z
+		return nil
+	}
+}
+
+func (a *ExpressionUnmarshaler) UnmarshalJSON(b []byte) error {
+	var expressionTypeName string
+	err := json.Unmarshal(b, &struct{ Type *string }{&expressionTypeName})
+	if err != nil {
+		return err
+	}
+	if et, ok := expressionTypeMap[expressionTypeName]; ok {
+		*a.Expression = et()
+		return json.Unmarshal(b, &a.Expression)
+	}
+	return fmt.Errorf("unknown Expression type")
+}
+
+func (a *ExpressionsUnmarshaler) UnmarshalJSON(b []byte) error {
+	var rms []json.RawMessage
+	err := json.Unmarshal(b, &rms)
+	if err != nil {
+		return err
+	}
+	*a.Expressions = make([]Expression, len(rms))
+	for i, v := range rms {
+		err := json.Unmarshal(v, &ExpressionUnmarshaler{&(*a.Expressions)[i]})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *LocationUnmarshaler) UnmarshalJSON(b []byte) error {
+	var locationKind string
+	var stringLocation common.StringLocation = ""
+	err := json.Unmarshal(b, &struct {
+		Type   *string
+		String *common.StringLocation
+	}{Type: &locationKind, String: &stringLocation})
+	if err != nil {
+		return err
+	}
+	if stringLocation != "" {
+		*a.Location = stringLocation
+		return nil
+	}
+	if f, ok := locationKindMap[locationKind]; ok {
+		*a.Location = f()
+		return nil
+	}
+	return fmt.Errorf("unknown Location")
+}
+
+func (a *StatementUnmarshaler) UnmarshalJSON(b []byte) error {
+	var statementKindName string
+	if err := json.Unmarshal(b, &struct{ Type *string }{&statementKindName}); err != nil {
+		return err
+	}
+	if sk, ok := statementKindMap[statementKindName]; ok {
+		*a.Statement = sk()
+		return json.Unmarshal(b, &a.Statement)
+	}
+	return fmt.Errorf("unknown Statement Kind")
+}
+
+func (a *StatementsUnmarshaler) UnmarshalJSON(b []byte) error {
+	var rms []json.RawMessage
+	err := json.Unmarshal(b, &rms)
+	if err != nil {
+		return err
+	}
+	*a.Statements = make([]Statement, len(rms))
+	for i, v := range rms {
+		err := json.Unmarshal(v, &StatementUnmarshaler{&(*a.Statements)[i]})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *TypeUnmarshaler) UnmarshalJSON(b []byte) error {
+	var typeName string
+	err := json.Unmarshal(b, &struct{ Type *string }{&typeName})
+	if err != nil {
+		return err
+	}
+	if f, ok := typeNameMap[typeName]; ok {
+		*a.Type = f()
+		return json.Unmarshal(b, &a.Type)
+	}
+	return fmt.Errorf("unknown Type")
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+func (a *Access) UnmarshalJSON(b []byte) error {
+	if c, ok := accessKindMap[strings.Trim(string(b), `"`)]; ok {
```

Comment:
Instead of trimming the serialized string's quotes manually, potentially leaving quoted content, unmarshal as a string:

```go
var s string
if err := json.Unmarshal(b, &s); err != nil {
	return err
}
```